### PR TITLE
[SuperTextField][Mobile] Always compute the viewport height in the first frame (Resolves #939)

### LINKED ISSUE #939: [SuperTextField] flicker on initial display (iOS)
In the video below we replace Text widges with `SuperTextField` upon entering edit mode. You can see that the text is missing for 1-2 frames. This does not happen on desktop.

https://user-images.githubusercontent.com/96958/212548613-c5fd706d-72eb-4aa8-86bf-14699d6927b4.mov

The culprit seems to be `TextScrollView.build`:

https://github.com/superlistapp/super_editor/blob/73990b7a72675aa730840b095e458dcde0c778f6/super_editor/lib/src/infrastructure/super_textfield/infrastructure/text_scrollview.dart#L466-L476

The layout is not calculated synchronously, so for several frames there is no content. But in this particular case calling `_updateViewportHeight();` directly from the build method seem to work without issues and resolves the flicker.

Is the asynchronous layout necessary? If so, can we determine cases where it's not a do it synchronously?

### PR DESCRIPTION
**Pull request by angelosilvestre**
[SuperTextField][Mobile] Always compute the viewport height in the first frame. Resolves #939

Currently, there are cases where we compute the textfield size after the first frame. This causes some flickering when using it inside a scrollable.

This PR moves the existing viewport height computation to a new widget, which performs the computation during layout. With this change, we always size the viewport since the first frame.

The downside is that we perform layout twice. The first one is needed so the text layout is performed so we can access the text layout information and the second one uses the information retrieved from the first layout to size the viewport. We probably could introduce some caching to avoid performing two layouts in some situations.

### CODE REVIEW

**Comment by matthew-carroll** [comment]
> Currently, there are cases where we compute the textfield size after the first frame. This causes some flickering when using it inside a scrollable.

Can you please provide a deeper description of the situation? In which cases do we do that, and why? And do you mean that we compute the size on the first frame AND after the first frame? Or are you saying that we don't compute it on the first frame, and instead we compute the initial size in a later frame?

**Comment by angelosilvestre** [comment]
When we have `minLines` or `maxLines`, we might need to set the textfield size to a fixed size in the following situations:

- If `minLines != null`, we check if the current text height is smaller than the amount of pixels of `minLines` of text. If it is, we set the textfield size to be `minLines * lineHeight`.
- If `maxLines != null`, we check if the current text height is greater than the amount of pixels of `maxLines` of text. If it is, we set the textfield size to be `maxLines * lineHeight`. 

Padding was omitted for simplicity.

To be able to measure the height of the text, text must be laid out first. In the first `build` call the text still isn't laid out, so we defer to the end of the frame.

The goal of this PR is to remove this computations from `TextScroller`, so during the build phase, it doesn't matter if the text is laid out or not.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 1]
```diff
@@ -338,191 +330,10 @@ class _TextScrollViewState extends State<TextScrollView>
...
-      return 0;
-    }
-
-    if (widget.textKey.currentState == null) {
-      return null;
-    }
-
-    return _textLayout.getLineCount();
+  void _scheduleRebuild() {
+    setState(() {
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:334]: How do we know that `textEditingController` won't call this method during a build? If it does, this `setState()` is going to blow up.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 2]
```diff
@@ -1095,24 +897,260 @@ enum _AutoScrollDirection {
...
-/// or unbounded, which means it should expand to fit its content.
-class _ViewportHeight {
-  const _ViewportHeight({
-    this.value,
+/// If the [child]'s desired height is smaller than [minLines], the size is set to acommodate [minLines] of
+/// text, plus [padding].
+///
+/// If the [child]'s desired height is bigger than [maxLines], the size is set to acommodate [maxLines] of
+/// text, plus [padding].
+class _TextViewport extends SingleChildRenderObjectWidget {
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:909]: A "viewport" means the bounds of the visible region of a scrolling experience. This doesn't seem to be a viewport?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 3]
```diff
@@ -1095,24 +897,260 @@ enum _AutoScrollDirection {
   end,
 }
 
-/// The height of the viewport.
+/// Sizes a [child] containing a [SuperText] to respect the given [minLines] and [maxLines].
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:900]: I'm confused by a lot of references `[child]` in this vs the concept of a descendent `SuperText`. If this render object truly handles any subtree that happens to contain a `SuperText` somewhere in it, then surely there's more to consider than just the text height and the padding, right? On the other hand, if the `child` is expected to be a `SuperText` then that changes how these docs should be written.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 4]
```diff
@@ -1095,24 +897,260 @@ enum _AutoScrollDirection {
...
+    required this.textKey,
+    this.minLines,
+    this.maxLines,
+    this.lineHeight,
+    this.padding,
+    super.child,
   });
 
-  /// Whether or not the viewport height is bounded.
+  /// The text that the [child] will display.
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:920]: If the child contains this text, then shouldn't this render object base itself on teh child's layout size? This render object seems to be replicating a bunch of implementation details of text layout, which the `child` should already be doing on our behalf.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 5]
```diff
@@ -1095,24 +897,260 @@ enum _AutoScrollDirection {
...
+      return;
+    }
+
+    _padding = value;
+    markNeedsLayout();
+  }
+
+  @override
+  void performLayout() {
+    if (child == null) {
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1047]: Is there any reasonable use of this render object where no child would be provided? If not, let's make it required and not worry about thi.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 6]
```diff
@@ -1095,24 +897,260 @@ enum _AutoScrollDirection {
...
+      size = Size.zero;
+      return;
+    }
+
+    // Perform a first layout so we can access information about the text layout.
+    child!.layout(constraints, parentUsesSize: true);
+
+    final viewportHeight = _computeViewportHeight();
+    if (viewportHeight == null) {
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1056]: I'm not a fan of using a `null` value for this. It's very strange to read. Why would our height be null? Something should probably change here with regard to naming or responsibility breakdown.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 7]
```diff
@@ -531,29 +342,20 @@ class _TextScrollViewState extends State<TextScrollView>
...
-        height: _viewportHeight,
-        child: Stack(
-          children: [
-            _buildScrollView(
-              child: widget.child,
-            ),
-            if (widget.showDebugPaint) ..._buildDebugScrollRegions(),
-          ],
-        ),
+    return _TextViewport(
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:345]: There's no great place to leave this comment, so I'll put it here. Can you write a test that demonstrates the desired effect? It should fail for the current case where we need to shrink, and another test where we need to expand. Then it should pass with these changes.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 4 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:920]: The child is a subtree like `Stack`-> `SingleChildScrollView`-> Text. When laying out the child without aditional constraints, it's taking all the available height.

Comment (matthew-carroll) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:920]: What if you surround `SingleChildScrollView` with an `IntrinsicHeight` widget? Or, if this needs to be done imperatively during layout, what if you `computeIntrinsicHeight()` on the `child`?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 2 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:909]: Updated.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 3 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:900]: Updated.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 4 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:920]: Updated.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 6 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1056]: Updated.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 7 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:345]: Added tests.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 8]
```diff
@@ -1,20 +1,20 @@
 import 'dart:math';
 
 import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
 import 'package:flutter/scheduler.dart';
-import 'package:super_editor/src/infrastructure/_logging.dart';
 import 'package:super_editor/src/infrastructure/flutter/flutter_pipeline.dart';
-import 'package:super_editor/src/super_textfield/super_textfield.dart';
+import 'package:super_editor/super_editor.dart';
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:7]: A piece of `super_editor` shouldn't import `super_editor`.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 9]
```diff
@@ -129,7 +126,7 @@ class _TextScrollViewState extends State<TextScrollView>
       ..delegate = this
       ..addListener(_onTextScrollChange);
 
-    widget.textEditingController.addListener(_scheduleViewportHeightUpdateAndRebuild);
+    widget.textEditingController.addListener(scheduleBuildAfterBuild);
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:129]: Even though you can call this method directly, doing so removes our opportunity to explain why we're doing that. Which means no one will risk touching it again. Can you add back a method and include an explanation in that method for why we need to schedule another build after every text editing change?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 10]
```diff
@@ -147,17 +144,13 @@ class _TextScrollViewState extends State<TextScrollView>
...
+      oldWidget.textEditingController.removeListener(scheduleBuildAfterBuild);
+      widget.textEditingController.addListener(scheduleBuildAfterBuild);
     }
 
     if (widget.minLines != oldWidget.minLines ||
         widget.maxLines != oldWidget.maxLines ||
-        widget.lineHeight != oldWidget.lineHeight) {
-      _updateViewportHeight();
-    }
+        widget.lineHeight != oldWidget.lineHeight) {}
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:153]: This is an empty if-statement.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 11]
```diff
@@ -338,222 +331,26 @@ class _TextScrollViewState extends State<TextScrollView>
...
-        child: Stack(
-          children: [
-            _buildScrollView(
-              child: widget.child,
-            ),
-            if (widget.showDebugPaint) ..._buildDebugScrollRegions(),
-          ],
-        ),
+    return _TextLinesLimiter(
+      text: widget.textEditingController.text,
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:341]: It looks like we're still depending on text from this new widget. Our discussion about intrinsic height was intended to remove text-specific knowledge - especially outside a scrollview. The other conversation just had an "Updated" response. Can you please let me know what you're thinking at this point with this approach?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 12]
```diff
@@ -1,20 +1,20 @@
...
+import 'package:super_editor/super_editor.dart';
 import 'package:super_text_layout/super_text_layout.dart';
 
 final _log = scrollingTextFieldLog;
 
 /// A scrollable that positions its [child] based on text metrics.
 ///
-/// The [child] must contain a [SuperSelectableText] in its tree,
-/// [textKey] must refer to that [SuperSelectableText], and the
+/// The [child] must contain a [SuperTextWithSelection] in its tree,
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:14]: Was this a correction of an old name, or did we actually change the requirements?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 11 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:341]: We don't need the `text` property anymore, I forgot to remove it. However, we still need to access the text layout to be able to query the line height so we are able to determine how many pixels `minLines` and `maxLines` of text are.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 12 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:14]: A correction of an old name.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 13]
```diff
@@ -1095,24 +894,223 @@ enum _AutoScrollDirection {
   end,
 }
 
-/// The height of the viewport.
+/// Sizes a [child] to respect the given [minLines] and [maxLines].
+///
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:898]: Let's add a line here:

```
/// The height of the [child] is calculated as the [child]'s `intrinsicHeight`, plus the 
/// given [padding] above and below the [child].
```

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 14]
```diff
@@ -1095,24 +894,223 @@ enum _AutoScrollDirection {
...
+
+    // Perform a first layout so we can access information about the text layout.
+    child!.layout(constraints, parentUsesSize: true);
+
+    final lineHeight = _computeLineHeight();
+    final minHeight = _computeMinHeight(lineHeight);
+    final maxHeight = _computeMaxHeight(lineHeight);
+
+    // The height we need to enforce if the child doesn't already respects the line restrictions.
+    double? enforcedHeight;
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1046]: `enforcedHeight` -> `adjustedChildHeight`

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 15]
```diff
@@ -1095,24 +894,223 @@ enum _AutoScrollDirection {
...
+
+    final childIntrinsicHeight = child!.getMinIntrinsicHeight(constraints.maxWidth);
+    if (childIntrinsicHeight <= minHeight) {
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1049]: Should this `<=` be `<`?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 16]
```diff
@@ -1095,24 +894,223 @@ enum _AutoScrollDirection {
...
+    );
+
+    // Perform the second layout with the adjusted constraints.
+    child!.layout(childConstraints, parentUsesSize: true);
+
+    size = child!.size;
+  }
+
+  double _computeMinHeight(double lineHeight) {
+    final totalVerticalPadding = _padding?.vertical ?? 0.0;
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1077]: Does the word `total` mean something here? This looks like just a normal `verticalPadding`

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 17]
```diff
@@ -1095,24 +894,223 @@ enum _AutoScrollDirection {
...
+  EdgeInsets? _padding;
+  set padding(EdgeInsets? value) {
+    if (value == _padding) {
+      return;
+    }
+
+    _padding = value;
+    markNeedsLayout();
+  }
+
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1027]: This render object extends `RenderProxyBox`, so it already has the implementations and the mixins.

File: super_editor/test/super_textfield/super_textfield_rendering_test.dart [hunk 18]
```diff
@@ -31,6 +32,68 @@ void main() {
...
+        controller: AttributedTextEditingController(
+          text: AttributedText('1'),
+        ),
+        showTextField: showTextField,
+        minLines: 5,
+      );
+
+      // Switch to display the SuperTextField.
+      // As our text has only one line, we should expand the text field height
+      // to acommodate 5 lines.
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor/test/super_textfield/super_textfield_rendering_test.dart:52]: I think that `// As our text has only one line, we should expand the text field height to 5 lines.` should be enough to communicate what's happening. What do you think?

File: super_editor/test/super_textfield/super_textfield_rendering_test.dart [hunk 19]
```diff
@@ -31,6 +32,68 @@ void main() {
...
+        tester,
+        controller: AttributedTextEditingController(
+          text: AttributedText('1'),
+        ),
+        showTextField: showTextField,
+        minLines: 5,
+      );
+
+      // Switch to display the SuperTextField.
+      // As our text has only one line, we should expand the text field height
```
[the remark this answers is not in the document]
Comment (angelosilvestre) [reply super_editor/test/super_textfield/super_textfield_rendering_test.dart:51]: > I'm guessing that the reason you're flipping this value is to show the text field and inspect that text field on its very first frame.

Yeah, I will include this information.

**Comment by angelosilvestre** [comment]
@matthew-carroll 

After changing from  `if (childIntrinsicHeight <= minHeight)` to `if (childIntrinsicHeight < minHeight)` the following test failed:

```dart
testWidgetsOnAllPlatforms('recalculates its viewport height when text changes for text bigger than maxLines',
        (tester) async {
      final controller = AttributedTextEditingController();

      await tester.pumpWidget(
        _buildScaffold(
          child: SuperTextField(
            minLines: 1,
            maxLines: 2,
            textController: controller,
          ),
        ),
      );
      await tester.pumpAndSettle();

      final textFieldSizeBefore = tester.getSize(find.byType(SuperTextField));

      // Change the text, so the content height is greater
      // than the initial content height.
      controller.text = AttributedText(
        """
This is
a
multi-line
SuperTextField
""",
      );
      await tester.pumpAndSettle();

      // Ensure the text field height has increased.
      expect(tester.getSize(find.byType(SuperTextField)).height, greaterThan(textFieldSizeBefore.height));
    });
```

As the text is initially empty, the `child` intrinsic height was already between `minLines` and `maxLines`, so we don't enforced any specific height. Because of that, the `SingleChildScrollView` expanded to fill all the available height.

I modified the implementation so that, if the `child`'s intrinsic size already respects `minLines` and `maxLines`, we force it to be exactly the height of its intrinsic height.

**Review by matthew-carroll — APPROVED** [review]
LGTM

**Pull request merged**